lib/IDE/Refactoring.cpp

```diff
@@ -7513,12 +7591,31 @@ class AsyncConverter : private SourceEntityWalker {
     OS << tok::l_brace;
   }
 
+  enum class PlaceholderMode {
+    SUCCESS_BLOCK, ERROR_BLOCK, FALLBACK
+  };
+
   void preparePlaceholdersAndUnwraps(AsyncHandlerDesc HandlerDesc,
-                                     ArrayRef<const ParamDecl *> SuccessParams,
-                                     const ParamDecl *ErrParam, bool Success) {
+                                     const ClosureCallbackParams &Params,
+                                     PlaceholderMode Mode) {
+    // Params that have been dropped always need placeholdering.
+    for (auto *Param : Params.getAllParams()) {
+      if (!Params.hasBinding(Param))
+        Placeholders.insert(Param);
+    }
+    // For the fallback case, no other params need placeholdering, as they are
+    // all freely accessible in the fallback case.
+    if (Mode == PlaceholderMode::FALLBACK)
+      return;
```

Comment:
I mention this in the tests below, but IMO we shouldn't placeholder the bool param either. As mentioned there though, happy to do that later.